core/src/main/java/com/datastrato/gravitino/storage/relational/JDBCBackend.java

```diff
@@ -0,0 +1,180 @@
+
+import com.datastrato.gravitino.Config;
+import com.datastrato.gravitino.Configs;
+import com.datastrato.gravitino.Entity;
+import com.datastrato.gravitino.EntityAlreadyExistsException;
+import com.datastrato.gravitino.HasIdentifier;
+import com.datastrato.gravitino.NameIdentifier;
+import com.datastrato.gravitino.Namespace;
+import com.datastrato.gravitino.exceptions.AlreadyExistsException;
+import com.datastrato.gravitino.exceptions.NoSuchEntityException;
+import com.datastrato.gravitino.meta.BaseMetalake;
+import com.datastrato.gravitino.storage.relational.mapper.MetalakeMetaMapper;
+import com.datastrato.gravitino.storage.relational.po.MetalakePO;
+import com.datastrato.gravitino.storage.relational.session.SqlSessionFactoryHelper;
+import com.datastrato.gravitino.storage.relational.utils.POConverters;
+import com.datastrato.gravitino.storage.relational.utils.SessionUtils;
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * {@link JDBCBackend} is a jdbc implementation of {@link RelationalBackend} interface. You can use
+ * a database that supports the JDBC protocol as storage. If the specified database has special SQL
+ * syntax, please implement the SQL statements and methods in MyBatis Mapper separately and switch
+ * according to the {@link Configs#ENTITY_RELATIONAL_JDBC_BACKEND_DB_TYPE_KEY} parameter.
+ */
+public class JDBCBackend implements RelationalBackend {
+
+  /** Initialize the jdbc backend instance. */
+  @Override
+  public void initialize(Config config) {
+    SqlSessionFactoryHelper.getInstance().init(config);
+  }
+
+  @Override
+  public <E extends Entity & HasIdentifier> List<E> list(
+      Namespace namespace, Entity.EntityType entityType) {
+    if (entityType == Entity.EntityType.METALAKE) {
+      List<MetalakePO> metalakePOS =
+          SessionUtils.getWithoutCommit(
+              MetalakeMetaMapper.class, MetalakeMetaMapper::listMetalakePOs);
+      return (List<E>) POConverters.fromMetalakePOs(metalakePOS);
+    } else {
+      throw new IllegalArgumentException(
+          String.format("Unsupported entity type: %s for list operation", entityType));
+    }
+  }
+
+  @Override
+  public boolean exists(NameIdentifier ident, Entity.EntityType entityType) {
+    if (entityType == Entity.EntityType.METALAKE) {
+      MetalakePO metalakePO =
+          SessionUtils.getWithoutCommit(
+              MetalakeMetaMapper.class, mapper -> mapper.selectMetalakeMetaByName(ident.name()));
+      return metalakePO != null;
+    } else {
+      throw new IllegalArgumentException(
+          String.format("Unsupported entity type: %s for exists operation", entityType));
+    }
+  }
+
+  @Override
+  public <E extends Entity & HasIdentifier> void insert(E e, boolean overwritten)
+      throws EntityAlreadyExistsException {
+    if (e instanceof BaseMetalake) {
+      MetalakePO metalakePO =
```

Comment:
This logic seems weird. But I am not familiar with JDBC. I don't know whether this is a good practice. Does JDBC have some standard error code to indicate exist exception?

Reply:
Here, I refer to the logic in KV Store for implementation. Currently, MyBatis handles exists returns the PersistenceException. This is a general exception. We need to handle it based on the message. I think it may not be a good way.

Reply:
I may have similar code to use the message of exception to throw another exception, you can search the code of the project.

Reply:
This operation isn't atomic. What will it happen if we insert a record after condition judgement in another thread?

Reply:
MySQL's `primary key` and `unique key` guarantee the atomicity. The judgment here actually only affects the case where `overwrite` is set to `false`, because when `overwrite` is set to `true`, we always use the overwrite strategy. For the case where `overwrite` is `false`, when thread `A` detects that there is no existing data `(id=1, metalake_name='test', metalake_comment='this is test')`, another thread `B` inserts new data `(id=1, metalake_name='test', metalake_comment='this is test')` . When thread `A` initiates an insert SQL, since there is no `overwriting` semantics, an exception will be thrown due to a conflict with the `primary key` (id) and the `unique key` (metalake_name), which will not affect the correctness of the data.

Reply:
> MySQL's `primary key` and `unique key` guarantee the atomicity. The judgment here actually only affects the case where `overwrite` is set to `false`, because when `overwrite` is set to `true`, we always use the overwrite strategy. For the case where `overwrite` is `false`, when thread `A` detects that there is no existing data `(id=1, metalake_name='test', metalake_comment='this is test')`, another thread `B` inserts new data `(id=1, metalake_name='test', metalake_comment='this is test')` . When thread `A` initiates an insert SQL, since there is no `overwriting` semantics, an exception will be thrown due to a conflict with the `primary key` (id) and the `unique key` (metalake_name), which will not affect the correctness of the data.

Will it throw a correct exception information?

Reply:
> > MySQL's `primary key` and `unique key` guarantee the atomicity. The judgment here actually only affects the case where `overwrite` is set to `false`, because when `overwrite` is set to `true`, we always use the overwrite strategy. For the case where `overwrite` is `false`, when thread `A` detects that there is no existing data `(id=1, metalake_name='test', metalake_comment='this is test')`, another thread `B` inserts new data `(id=1, metalake_name='test', metalake_comment='this is test')` . When thread `A` initiates an insert SQL, since there is no `overwriting` semantics, an exception will be thrown due to a conflict with the `primary key` (id) and the `unique key` (metalake_name), which will not affect the correctness of the data.
> 
> Will it throw a correct exception information?

In response to the above situation, a unique constraint exception will generally be thrown. The JDBC protocol provides an exception interface, but different databases may have different primary key and unique key conflict exception implementations, and we may need to handle the return messages of different databases.

Reply:
A simple way to handle it is to consider it as existing when we catch the `SQLIntegrityConstraintViolationException`, because this exception usually means that SQL violates the constraints of the `primary key` and `unique key`.

Reply:
Does JDBC have a standard error code or exception?

Reply:
> A simple way to handle it is to consider it as existing when we catch the `SQLIntegrityConstraintViolationException`, because this exception usually means that SQL violates the constraints of the `primary key` and `unique key`.

@qqqttt123 The `SQLIntegrityConstraintViolationException` I mentioned is the standard exception class of Java SQL. Other databases that support JDBC need to be extended based on this class.

Reply:
OK, Maybe we should provide some interfaces to handle the the mapping exceptions.
NotExistException ExistException and so on.